File: data_structures/linked_list/linked_list.go
package main
import "fmt"

type Node struct {
  value int
  next *Node
}

func NewNode(value int, next *Node) *Node {
  node := new(Node)
  node.value = value
  node.next = next

  return node
}

type List struct {
  root *Node
}

func (l *List) add(value int) {
  node := NewNode(value, nil)

  if (l.root == nil) {
    l.root = node
  } else {
    l._add(l.root, node)
  }
}

func (l *List) find(value int) {
  isInList := "is NOT"

  if (l._find(l.root, value)) {
    isInList = "is"
  }

  fmt.Println("Value", value, isInList, "in list")
}

func (l *List) print() {
  l._print(l.root)
}

func (l *List) _add(root *Node, node *Node) {
  if (root.value < node.value &&
      (root.next == nil || root.next.value > node.value)) {
    temp := root.next
    root.next = node
    node.next = temp
  } else {
    l._add(root.next, node)
  }
}

func (l *List) _find(root *Node, value int) bool {
  if (root == nil) {
    return false
  } else if (root.value == value) {
    return true
  } else {
    return l._find(root.next, value)
  }
}

func (l *List) _print(root *Node) {
  if (root != nil) {
    fmt.Println(root.value)
    l._print(root.next)
  }
}

func main() {
  list := List{nil}

  list.add(1)
  list.add(2)
  list.add(3)
  list.print()

  list.find(3)
  list.find(4)
}

File: data_structures/linked_list/linked_list.cpp
using namespace std;
#include <iostream>
#include <string>

/**
 * Defines a node in a Linked List.
 *
 * Only supports integer values.
 */
class Node {
  public:
    int value;
    Node* next;

    Node(int value, Node* next=NULL) {
      this->value = value;
      this->next = next;
    }
};

/**
 * Defines a sorted, singly Linked List.
 *
 * Any values added will be added in order.
 * This class also provides search and deleteion functionality.
 */
class List {
  public:
    Node* root;

    List(Node* root=NULL) {
      this->root = root;
    }

    void add(int value) {
      Node* node = new Node(value);

      // if the list is empty, create a new root.
      // Otherwise, call private `_add` method.
      if (!this->root) {
        this->root = node;
      } else {
        this->_add(this->root, node);
      }
    }

    void find(int value) {
      string isInList = "is NOT";

      if (this->_find(this->root, value)) {
        isInList = "is";
      }

      cout << "Value " << value << " " << isInList << " in list" << endl;
    }

    void print() {
      this->_print(this->root);
    }

  private:
    void _add(Node* root, Node* node) {
      // check if `next` is null (end of list)
      // check if `node->value` falls between `root` and `root-next` values.
      if (root->value < node->value && (
          !root->next || root->next->value > node->value)) {
        // preform an addition between two existing nodes or at the end of list.
        Node* oldNext = root->next;
        root->next = node;
        node->next = oldNext;
      } else {
        this->_add(root->next, node);
      }
    }

    bool _find(Node* root, int value) {
      if (!root) {
        return false;
      } else if (root->value == value) {
        return true;
      } else {
        return this->_find(root->next, value);
      }
    }

    void _print(Node* root) {
      if (root) {
        cout << root->value << endl;
        this->_print(root->next);
      }
    }
};

int main() {

  List* list = new List();

  list->add(1);
  list->add(3);
  list->add(2);

  list->print();
  list->find(3);
  list->find(4);

  return 0;
};
